### PR TITLE
mcpelauncher-{client,ui-qt}: init at 1.1.2

### LINKED ISSUE #231875: Package request: mcpelauncher
**Project description**

An unofficial launcher for Minecraft Bedrock Edition

**Metadata**

* homepage URL: [https://mcpelauncher.readthedocs.io](https://mcpelauncher.readthedocs.io)
* source URL: [https://github.com/minecraft-linux/mcpelauncher-manifest](https://github.com/minecraft-linux/mcpelauncher-manifest)
* license: GPL 3.0
* platforms: Linux and Darwin

### PR DESCRIPTION
**Pull request by Morxemplum**
Closes #231875.

This is a second attempt at introducing mcpelauncher to nixpkgs, after the previous attempt (#232256) wasn't merged as work stopped being done on it.

I was able to take @Aleksanaa's changes and freshen them up to the latest version, fixing the issues she ran into along the way.

### Main changes since previous pull request
- Brought the package up to the latest release (1.1.2)
- GLFW is no longer optional, and SDL is completely abandoned in this package
    - GLFW is now the main game-window frontend
    - Ensure that Nix's GLFW is used
- Added native wayland support
    - Fixed some issues with relative scaling in regards to fractional scaling
- Fixed the audio issues present by re-enabling an experimental PulseAudio backend
- Upgraded from Qt5 to Qt6

### badPlatforms = lib.platforms.darwin

Minecraft is planning on [updating their graphics minimum requirements](https://help.minecraft.net/hc/en-us/articles/30298767427597-Upcoming-OS-Sunset-Announcements-in-Minecraft) to OpenGL ES 3.1, something that the upstream package [does not support for macOS](https://github.com/minecraft-linux/mcpelauncher-manifest/issues/1042#issuecomment-2367388410) as it only does OpenGL ES 3.0 through emulation. This change takes effect starting March 2025. I don't know if the maintainer has any plans on upgrading to keep supporting macOS, but for right now I am deciding to not worry about darwin support.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/mc/mcpelauncher-client/nlohmann_json_373_dep.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-client/nlohmann_json_373_dep.nix:1]: Can't we just override nlohmann_json?

File: pkgs/by-name/mc/mcpelauncher-client/package.nix [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      "-DBUILD_WEBVIEW=OFF"
+      "-DXAL_WEBVIEW_USE_CLI=ON"
+      "-DXAL_WEBVIEW_USE_QT=OFF"
+    ]
+    ++ lib.optionals (!withQtErrorWindow) [
+      "-DENABLE_QT_ERROR_UI=OFF"
+    ];
+
+  meta = {
+    description = "Unofficial *NIX launcher of Minecraft: Bedrock Edition using Android - CLI launcher";
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-client/package.nix:107]: Feels a bit wordy IMO

```suggestion
    description = "Unofficial *NIX launcher of Minecraft: Bedrock Edition";
```

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: pkgs/by-name/mc/mcpelauncher-client/package.nix [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  };
+
+  cmakeFlags =
+    [
+      "-DUSE_OWN_CURL=OFF"
+      "-DENABLE_DEV_PATHS=OFF"
+      "-Wno-dev"
+      "-DCMAKE_BUILD_TYPE=Release"
+      "-DGAMEWINDOW_SYSTEM=GLFW"
+      "-DUSE_SDL3_AUDIO=OFF"
```
[the remark this answers is not in the document]
Comment (Morxemplum) [reply pkgs/by-name/mc/mcpelauncher-client/package.nix:109]: Interesting, This must be a guideline that was adopted in the time between the previous pull request and this one.

File: pkgs/by-name/mc/mcpelauncher-ui-qt/package.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    protobuf
+    qt6.qtwebengine
+    qt6.qtsvg
+    qt6.qtwayland
+    glfw
+  ];
+
+  cmakeFlags = [
+    "-Wno-dev"
+    "-DCMAKE_BUILD_TYPE=Release"
```
[the remark this answers is not in the document]
Comment (Morxemplum) [reply pkgs/by-name/mc/mcpelauncher-ui-qt/package.nix:51]: My bad. I set this flag early on when I was basically trying to figure out why the project wasn't building and wanted to make sure that I was following the instructions to a T. I can remove this in some clean up.

**Comment by Morxemplum** [comment]
Here is some further showcase of the client and the Qt UI at work. Attached below are screenshots, and a full showcase video can be seen [here](https://youtu.be/DUt0OA5IqPk).

60841367-fde0-4eb9-96c4-1972d23375bb

Screenshot From 2024-12-14 14-26-39

Screenshot From 2024-12-14 14-27-00

Screenshot From 2024-12-14 22-27-50

Screenshot From 2024-12-14 22-28-16

File: pkgs/by-name/mc/mcpelauncher-client/dont_download_glfw_client.patch [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
++++ b/ext/glfw.cmake
+@@ -1,18 +1,2 @@
+-include(FetchContent)
+-
+-set(GLFW_BUILD_DOCS OFF CACHE BOOL "" FORCE)
+-set(GLFW_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+-set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+-set(GLFW_INSTALL OFF CACHE BOOL "" FORCE)
+-
+-FetchContent_Declare(
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-client/dont_download_glfw_client.patch:13]: This patch should also not be necessary as it's also using FetchContent (you can probably get away with setting the source dir to `glfw.src`)

File: pkgs/by-name/mc/mcpelauncher-ui-qt/dont_download_glfw_ui.patch [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+@@ -1,19 +1,2 @@
+-include(FetchContent)
+-
+-set(GLFW_BUILD_DOCS OFF CACHE BOOL "" FORCE)
+-set(GLFW_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+-set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+-set(GLFW_INSTALL OFF CACHE BOOL "" FORCE)
+-set(GLFW_BUILD_WAYLAND OFF CACHE BOOL "" FORCE)
+-
+-FetchContent_Declare(
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-ui-qt/dont_download_glfw_ui.patch:13]: Ditto

File: pkgs/by-name/mc/mcpelauncher-client/package.nix [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        stripRoot = false;
+      })
+    ))
+    (lib.cmakeBool "USE_OWN_CURL" false)
+    (lib.cmakeBool "ENABLE_DEV_PATHS" false)
+    (lib.cmakeFeature "GAMEWINDOW_SYSTEM" "GLFW")
+    (lib.cmakeBool "USE_SDL3_AUDIO" false)
+    (lib.cmakeBool "BUILD_WEBVIEW" withQtWebview)
+    (lib.cmakeBool "XAL_WEBVIEW_USE_CLI" (!withQtWebview))
+    (lib.cmakeBool "XAL_WEBVIEW_USE_QT=OFF" withQtWebview)
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-client/package.nix:104]: Is this what you mean? Or something else?
```suggestion
    (lib.cmakeBool "XAL_WEBVIEW_USE_QT" false)
```

File: pkgs/by-name/mc/mcpelauncher-ui-qt/package.nix [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    qt6.qtwayland
+    glfw
+  ];
+
+  # the program refuses to start when QT_STYLE_OVERRIDE is set
+  # https://github.com/minecraft-linux/mcpelauncher-ui-qt/issues/25
+  preFixup = ''
+    qtWrapperArgs+=(
+      --prefix PATH : ${lib.makeBinPath [ mcpelauncher-client ]}
+      --set QT_STYLE_OVERRIDE ""
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-ui-qt/package.nix:54]:
```suggestion
      --unset QT_STYLE_OVERRIDE
```

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: pkgs/by-name/mc/mcpelauncher-client/dont_download_glfw_client.patch (hunk 5 above)
Comment (Aleksanaa) [reply pkgs/by-name/mc/mcpelauncher-client/dont_download_glfw_client.patch:13]: This is intentional as it avoids building glfw.

File: pkgs/by-name/mc/mcpelauncher-client/package.nix [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    description = "Unofficial Minecraft Bedrock Edition launcher with CLI";
+    homepage = "https://minecraft-linux.github.io";
+    license = lib.licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [
+      aleksana
+      morxemplum
+    ];
+    mainProgram = "mcpelauncher-client";
+    platforms = lib.platforms.unix;
+    badPlatforms = lib.platforms.darwin;
```
Comment (pluiedev) [inline pkgs/by-name/mc/mcpelauncher-client/package.nix:118]: Should probably also include the explanation as to why Darwin is marked broken here, like in the PR description

**Comment by pluiedev** [comment]
@ofborg eval

**Review by pluiedev — APPROVED** [review]

**Comment by Aleksanaa** [comment]
@ofborg build mcpelauncher-client mcpelauncher-ui-qt

**Comment by surfaceflinger** [comment]
Just in time to get the mojang office cape, thanks everyone!
